### ISSUE TITLE
Neo4j 3.4.5 - shortestPath with node label filters causes database freeze

### ISSUE THREAD

**Issue by vrazdobudko** [neo4j/neo4j#12016]
- Neo4j version: 3.4.5,
- Operating system: Ubuntu
- Hardware: 128 GB RAM, SSD, CPU Intel Xeon E5-2650

### Sysinfo
![default](https://user-images.githubusercontent.com/16487496/44140134-37db40ec-a08a-11e8-9cbd-b60251a22984.png)

### Сonfig
NEO4J_dbms_memory_heap_initial__size=30G
NEO4J_dbms_memory_heap_max__size=30G
NEO4J_dbms_memory_pagecache_size=64G
NEO4J_dbms_jvm_additional=-XX:NewRatio=1

### Query
```
OPTIONAL MATCH p = shortestPath((o1:Organization{})-[*1..30]-(o2:Organization{}))
WHERE o1.OrganizationId IN [6907471]
  AND o2.OrganizationId IN [507610]
  AND ALL (r IN relationships(p) WHERE (type(r) = 'OrganizationManagingCompany') OR (type(r) = 'OrganizationChief')
    OR (type(r) = 'OrganizationFounderOrganization') OR (type(r) = 'OrganizationFounderPerson'))
  AND ALL (n IN nodes(p) WHERE n:Person OR n:Organization) 
RETURN p
```

### Steps to reproduce
When I run the query on data which has no shortest path database freezes. Neo4j process still working with high cpu usage, but can't be accessible by http or bolt. And we can't get any metrics from jvm.

When I remove node label filters, query works properly.

**Query without node label filters:**
```
OPTIONAL MATCH p = shortestPath((o1:Organization{})-[*1..30]-(o2:Organization{}))
WHERE o1.OrganizationId IN [6907471]
  AND o2.OrganizationId IN [507610]
  AND ALL (r IN relationships(p) WHERE (type(r) = 'OrganizationManagingCompany') OR (type(r) = 'OrganizationChief')
    OR (type(r) = 'OrganizationFounderOrganization') OR (type(r) = 'OrganizationFounderPerson'))
RETURN p
```
**Query profile for data which has shortest path:**

1) With node label filters:
![default](https://user-images.githubusercontent.com/16487496/44140161-486880e6-a08a-11e8-9049-ee2a58adc667.png)

2) Without node label filters:
![default](https://user-images.githubusercontent.com/16487496/44140187-5c335d26-a08a-11e8-9a65-c3643c33dae7.png)

**Comment by SaschaPeukert** [comment]
Hej @vrazdobudko ,

Thanks for reporting this. We will have a look and get back to you.

**Comment by SaschaPeukert** [comment]
In the meantime, does the problem still exists with Neo4j 3.4.7?

**Comment by Lojjs** [comment]
@vrazdobudko The reason for the different plans with and without `ALL (n IN nodes(p) WHERE n:Person OR n:Organization) ` is explained in https://neo4j.com/docs/developer-manual/current/cypher/execution-plans/shortestpath-planning/ . The start node, end node and relationship predicates can be checked as we go during the search for the shortest path while the node predicate needs the entire path to evaluate the predicate. Therefore a fallback is planned according to section 3.7.5.2.

The fallback will make an exhaustive search for shortest paths, to avoid the scenario that the regular shortest path algorithm finds one path which will be thrown away in a later step due to not fulfilling the node predicate. Without the fallback your result would be that there where no shortest paths, even if there might have existed some longer path fulfilling all predicates. (Tell me if you need an example to understand this, it is hard to describe in text...).

It is probably this exhaustive fallback that freezes your large database without shortest paths. You could  try to either 

1. Set the config option `cypher.forbid_exhaustive_shortestpath` to true and see if the faster algorithm succeeds without the fallback or 

2. Change your query to 

```
OPTIONAL MATCH p = shortestPath((o1:Organization{})-[*1..30]-(o2:Organization{}))
WHERE o1.OrganizationId IN [6907471]
  AND o2.OrganizationId IN [507610]
  AND ALL (r IN relationships(p) WHERE (type(r) = 'OrganizationManagingCompany') OR (type(r) = 'OrganizationChief')
    OR (type(r) = 'OrganizationFounderOrganization') OR (type(r) = 'OrganizationFounderPerson'))
WITH p
 WHERE ALL (n IN nodes(p) WHERE n:Person OR n:Organization) 
RETURN p
```

which will make you end up in the case described in 3.7.5.3 instead. Do note the semantic difference though. Your query should return `the shortest path fulfilling ALL (n IN nodes(p) WHERE n:Person OR n:Organization) `, while mine returns `the shortest path IF it fulfils ALL (n IN nodes(p) WHERE n:Person OR n:Organization) `.

If neither of the 2 options above works for you, we can discuss further if there is something else we can do implementation wise.

Best regards Louise, Neo4j Cypher team

**Comment by Lojjs** [comment]
@vrazdobudko Since this isn't really a bug but more of a known performance limitation that can be enhanced (see my reply above), we have decided that we will not fix it in version 3.4. Also we are late in the release cycle for 3.5 but we have taken your feedback into account and an enhancement of shortest path that will allow your query to get a better plan will most probably come in the version after 3.5. Hope my advices above is to some help in the meantime.
